Remove cml_rt1011_rt5682 machine driver

brentlu opened the pull request:
This driver supports ALC5682 headphone codec with 2 or 4 ALC1011 speaker amplifier. Move the support to sof_rt5682 and sof_realtek_common module then delete the cml_rt1011_rt5682 machine driver. Tested on CML Chromebook Helios.

brentlu committed:
ASoC: Intel: sof_realtek_common: support 4xALC1011 amplifier

Add support for boards with four ALC1011 amplifiers. Configuration is
copied from cml_rt1011_rt5682 machine driver for backward
compatibility with existing cml devices.

Signed-off-by: Brent Lu <brent.lu@intel.com>

brentlu committed:
ASoC: Intel: sof_rt5682: support ALC1011 on cml boards

For cml boards, ALC1011 speaker amplifier is supported by machine
driver cml_rt1011_rt5682. Use same driver name for backward
compatibility with existing devices on market.

Signed-off-by: Brent Lu <brent.lu@intel.com>

brentlu committed:
ASoC: Intel: cml_rt1011_rt5682: delete driver

Delete this driver and use sof_rt5682 machine driver instead.

Signed-off-by: Brent Lu <brent.lu@intel.com>

plbossart commented on sound/soc/intel/boards/sof_realtek_common.c:
```diff
@@ -118,28 +172,98 @@ static const struct snd_soc_ops rt1011_ops = {
+		}
+
+		/* backward-compatible with existing cml devices */
+		fallthrough;
```
when there are 2 codecs we still deal with 4 speakers. is this intended?
how do we differentiate from the case with 4 codecs?

brentlu replied:
In this case num_codecs equals 2 so only first two speakers, "WL Ext Spk" and "WR Ext Spk", wil be installed.

plbossart commented on sound/soc/intel/boards/sof_realtek_common.c:
```diff
@@ -118,28 +172,98 @@ static const struct snd_soc_ops rt1011_ops = {
+	switch (num_codecs) {
+	case 2:
+		if (soc_intel_is_cml())
+			card->codec_conf = rt1011_4spk_codec_confs;
```
again a case with 2 codecs and 4 speakers, you should add a comment so that the new configuration is well identified. With the removal of the existing code we will lose context on why this was done this way.

brentlu replied:
will add a comment here; code here is just for compatibility with the implementation of cml_rt1011_rt5682 driver

plbossart commented on sound/soc/intel/boards/sof_rt5682.c:
```diff
@@ -639,6 +639,13 @@ static int sof_audio_probe(struct platform_device *pdev)
+		sof_audio_card_rt5682.name = devm_kstrdup(&pdev->dev,
+							  "cml_rt1011_rt5682",
+							  GFP_KERNEL);
+	}
```
test allocation success?

brentlu replied:
sure do

plbossart commented on sound/soc/intel/boards/Kconfig:
```diff
@@ -593,22 +593,6 @@ config SND_SOC_INTEL_CML_LP_DA7219_MAX98357A_MACH
-	  RT1011 + RT5682 I2S codec.
-	  Say Y if you have such a device.
-	  If unsure select "N".
-
```
if a .config file only selected this option, then with a make olddefconfig the sof_rt5682 machine would not be selected.
For true backwards compatibility I think we need to keep this Kconfig and use 'imply sof_rt5682' or something to keep the link.

brentlu replied:
will add the option back

plbossart commented:
@ujfalusi @marc-hb @fredoh9 any idea why the SOF kernel builds fail now?

It started last night, what changed?

fredoh9 commented:
SOFCI TEST

plbossart commented:
The CI tests report a failure on Helios, that's concerning.
https://sof-ci.01.org/linuxpr/PR4897/build2042/devicetest/index.html?model=CML_HEL_RT5682-ipc3&testcase=verify-kernel-boot-log

plbossart commented:
SOFCI TEST

marc-hb commented:
CML FW failing to load again in https://sof-ci.01.org/linuxpr/PR4897/build2055/devicetest/index.html so this was not a fluke.

marc-hb commented:
We have a well-known bug where we don't get the kernel logs when the firmware does not load. It's very painful. Please upvote and comment to help prioritize it:
https://github.com/thesofproject/sof-test/issues/1112

I used ssh and I fetch some logs, see below. UTC time (as it always should be)

```
Apr 04 14:39:13 jf-cml-hel-rt5682-07 kernel: snd_sof:sof_widget_ready: sof-audio-pci-intel-cnl 0000:00:1f.3: tplg: widget 0 (PCM0P) is ready [type: 23, pipe: 1, pins: 0 / 0, stream: Passthrough Playback 0]
Apr 04 14:39:13 jf-cml-hel-rt5682-07 kernel: snd_sof:sof_control_load: sof-audio-pci-intel-cnl 0000:00:1f.3: tplg: load control type 1 name : 1 Master Playback Volume
Apr 04 14:39:13 jf-cml-hel-rt5682-07 kernel: snd_sof:sof_control_load_volume: sof-audio-pci-intel-cnl 0000:00:1f.3: tplg: load kcontrol index 1 chans 2
Apr 04 14:39:13 jf-cml-hel-rt5682-07 kernel: snd_sof:sof_widget_ready: sof-audio-pci-intel-cnl 0000:00:1f.3: tplg: widget 1 (PGA1.0) is ready [type: 6, pipe: 1, pins: 0 / 0, stream: none]
Apr 04 14:39:13 jf-cml-hel-rt5682-07 kernel: snd_sof:sof_widget_ready: sof-audio-pci-intel-cnl 0000:00:1f.3: tplg: widget 2 (BUF1.0) is ready [type: 31, pipe: 1, pins: 0 / 0, stream: none]
Apr 04 14:39:13 jf-cml-hel-rt5682-07 kernel: snd_sof:sof_widget_ready: sof-audio-pci-intel-cnl 0000:00:1f.3: tplg: widget 3 (BUF1.1) is ready [type: 31, pipe: 1, pins: 0 / 0, stream: none]
Apr 04 14:39:13 jf-cml-hel-rt5682-07 kernel: snd_sof:sof_widget_ready: sof-audio-pci-intel-cnl 0000:00:1f.3: tplg: widget 4 (SSP0.OUT) is ready [type: 27, pipe: 1, pins: 0 / 0, stream: SSP0-Codec]
Apr 04 14:39:13 jf-cml-hel-rt5682-07 kernel: snd_sof:sof_connect_dai_widget: sof-audio-pci-intel-cnl 0000:00:1f.3: tplg: connected widget SSP0.OUT -> DAI link SSP0-Codec
Apr 04 14:39:13 jf-cml-hel-rt5682-07 kernel: snd_sof:sof_widget_ready: sof-audio-pci-intel-cnl 0000:00:1f.3: tplg: widget 5 (PIPELINE.1.SSP0.OUT) is ready [type: 32, pipe: 1, pins: 0 / 0, stream: SSP0.OUT]
Apr 04 14:39:13 jf-cml-hel-rt5682-07 kernel: snd_sof:sof_widget_ready: sof-audio-pci-intel-cnl 0000:00:1f.3: tplg: widget 6 (PCM0C) is ready [type: 24, pipe: 2, pins: 0 / 0, stream: Passthrough Capture 0]
Apr 04 14:39:13 jf-cml-hel-rt5682-07 kernel: snd_sof:sof_control_load: sof-audio-pci-intel-cnl 0000:00:1f.3: tplg: load control type 1 name : 2 Master Capture Volume
Apr 04 14:39:13 jf-cml-hel-rt5682-07 kernel: snd_sof:sof_control_load_volume: sof-audio-pci-intel-cnl 0000:00:1f.3: tplg: load kcontrol index 7 chans 2
Apr 04 14:39:13 jf-cml-hel-rt5682-07 kernel: snd_sof:sof_widget_ready: sof-audio-pci-intel-cnl 0000:00:1f.3: tplg: widget 7 (PGA2.0) is ready [type: 6, pipe: 2, pins: 0 / 0, stream: none]
Apr 04 14:39:13 jf-cml-hel-rt5682-07 kernel: snd_sof:sof_widget_ready: sof-audio-pci-intel-cnl 0000:00:1f.3: tplg: widget 8 (BUF2.0) is ready [type: 31, pipe: 2, pins: 0 / 0, stream: none]
Apr 04 14:39:13 jf-cml-hel-rt5682-07 kernel: snd_sof:sof_widget_ready: sof-audio-pci-intel-cnl 0000:00:1f.3: tplg: widget 9 (BUF2.1) is ready [type: 31, pipe: 2, pins: 0 / 0, stream: none]
Apr 04 14:39:13 jf-cml-hel-rt5682-07 kernel: snd_sof:sof_widget_ready: sof-audio-pci-intel-cnl 0000:00:1f.3: tplg: widget 10 (SSP0.IN) is ready [type: 28, pipe: 2, pins: 0 / 0, stream: SSP0-Codec]
Apr 04 14:39:13 jf-cml-hel-rt5682-07 kernel: snd_sof:sof_connect_dai_widget: sof-audio-pci-intel-cnl 0000:00:1f.3: tplg: connected widget SSP0.IN -> DAI link SSP0-Codec
Apr 04 14:39:13 jf-cml-hel-rt5682-07 kernel: snd_sof:sof_widget_ready: sof-audio-pci-intel-cnl 0000:00:1f.3: tplg: widget 11 (PIPELINE.2.SSP0.IN) is ready [type: 32, pipe: 2, pins: 0 / 0, stream: SSP0.IN]
Apr 04 14:39:13 jf-cml-hel-rt5682-07 kernel: snd_sof:sof_widget_ready: sof-audio-pci-intel-cnl 0000:00:1f.3: tplg: widget 12 (PCM1C) is ready [type: 24, pipe: 3, pins: 0 / 0, stream: Passthrough Capture 1]
Apr 04 14:39:13 jf-cml-hel-rt5682-07 kernel: snd_sof:sof_widget_ready: sof-audio-pci-intel-cnl 0000:00:1f.3: tplg: widget 13 (BUF3.0) is ready [type: 31, pipe: 3, pins: 0 / 0, stream: none]
Apr 04 14:39:13 jf-cml-hel-rt5682-07 kernel: snd_sof:sof_widget_ready: sof-audio-pci-intel-cnl 0000:00:1f.3: tplg: widget 14 (DMIC0.IN) is ready [type: 28, pipe: 3, pins: 0 / 0, stream: dmic01]
Apr 04 14:39:13 jf-cml-hel-rt5682-07 kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: error: can't connect DAI DMIC0.IN stream dmic01
Apr 04 14:39:13 jf-cml-hel-rt5682-07 kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: error: failed to add widget id 0 type 28 name : DMIC0.IN stream dmic01
Apr 04 14:39:13 jf-cml-hel-rt5682-07 kernel: sof_rt5682 cml_rt5682_def: ASoC: failed to load widget DMIC0.IN
Apr 04 14:39:13 jf-cml-hel-rt5682-07 kernel: sof_rt5682 cml_rt5682_def: ASoC: topology: could not load header: -22
Apr 04 14:39:13 jf-cml-hel-rt5682-07 kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: error: tplg component load failed -22
Apr 04 14:39:13 jf-cml-hel-rt5682-07 kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: error: failed to load DSP topology -22
Apr 04 14:39:13 jf-cml-hel-rt5682-07 kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: ASoC: error at snd_soc_component_probe on 0000:00:1f.3: -22
Apr 04 14:39:13 jf-cml-hel-rt5682-07 kernel: sof_rt5682 cml_rt5682_def: ASoC: failed to instantiate card -22
Apr 04 14:39:13 jf-cml-hel-rt5682-07 kernel: sof_rt5682 cml_rt5682_def: probe with driver sof_rt5682 failed with error -22
```

plbossart commented:
@brentlu looks like we have a disconnect between machine driver and topology now, so this PR can't be merged until CI tests are ok.

brentlu commented:
There is something wrong with the dmic dai link. May I know the configuration of the device? Is it running ChromeOS? What's the OS image version?

Commit:
ASoC: Intel: sof_rt5682: Fix uninitialized variable in probe

Initialize "is_legacy_cpu" to false to prevent an uninitialized variable
bug.

Fixes: 8efcd4864652 ("ASoC: Intel: sof_rt5682: use common module for sof_card_private initialization")
Signed-off-by: Dan Carpenter <dan.carpenter@linaro.org>
Acked-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.intel.com>
Link: https://msgid.link/r/6e16433f-1897-46a2-b3ee-e177e7d846b8@moroto.mountain
Signed-off-by: Mark Brown <broonie@kernel.org>

plbossart commented:
I can reproduce this issue manually on jf-cml-hel-rt5682-05, this really isn't a CI problem.

My guess is that we're now having a mismatch in topology, what's used by this PR is:

````
[    3.403953] snd_soc_sof_rt5682:sof_audio_probe: sof_rt5682 cml_rt5682_def: sof_rt5682_quirk = 801

[    3.263893] sof-audio-pci-intel-cnl 0000:00:1f.3:  Firmware file:     intel/sof/community/sof-cml.ri
[    3.263897] sof-audio-pci-intel-cnl 0000:00:1f.3:  Topology file:     intel/sof-tplg/sof-cml-rt1011-rt5682.tplg
````

[dmesg.log](https://github.com/thesofproject/linux/files/14910620/dmesg.log)

brentlu commented:
Something wrong with dmic01 dai link which does not happen on my side. Does the CI run a clean build since there is change in Makefile?

HW: helios
OS: hatch-release/R122-15708.0.0
kernel: chrome-v5.15
fw/tplg: inbox or fafc1ca00 (cavs2.5-001-drop-stable)

```
[    6.321469] sof-audio-pci-intel-cnl 0000:00:1f.3: tplg: ready widget id 12 pipe 3 type 24 name : PCM1C stream Passthrough Capture 1
[    6.321473] sof-audio-pci-intel-cnl 0000:00:1f.3: tplg: ready widget id 13 pipe 3 type 31 name : BUF3.0 stream none
[    6.321476] sof-audio-pci-intel-cnl 0000:00:1f.3: tplg: ready widget id 14 pipe 3 type 28 name : DMIC0.IN stream dmic01
[    6.321479] sof-audio-pci-intel-cnl 0000:00:1f.3: tplg: connected widget DMIC0.IN -> DAI link dmic01
[    6.321482] sof-audio-pci-intel-cnl 0000:00:1f.3: tplg: ready widget id 15 pipe 3 type 32 name : PIPELINE.3.DMIC0.IN stream DMIC0.IN
```

brentlu commented:
remove last patch to see if this PR still breaks the tplg loading

bardliao commented:
@brentlu @plbossart `dmic01` is defined in `cml_rt1011_rt5682_dailink[]` consistently, but, it depends on `ctx->dmic_be_num` in sof_rt5682.c. And I checked `jf-cml-hel-rt5682-05` the `ctx->dmic_be_num` value is 0.

Edit: is_legacy_cpu is true. That's why `ctx->dmic_be_num` is 0. And if I test dmic capture without this PR, I will get below error
```
[  175.731140] snd_sof:sof_ipc3_pcm_hw_params: sof-audio-pci-intel-cnl 0000:00:1f.3: stream_tag 2
[  175.731144] snd_sof:ipc3_log_header: sof-audio-pci-intel-cnl 0000:00:1f.3: ipc tx: 0x60010000: GLB_STREAM_MSG: PCM_PARAMS
[  175.731564] sof-audio-pci-intel-cnl 0000:00:1f.3: ipc tx error for 0x60010000 (msg/reply size: 108/20): -22
[  175.732173] sof-audio-pci-intel-cnl 0000:00:1f.3: HW params ipc failed for stream 2
[  175.732778] sof-audio-pci-intel-cnl 0000:00:1f.3: ASoC: error at snd_soc_pcm_component_hw_params on 0000:00:1f.3: -22
```
It let me wonder if CML_HEL_RT5682 really supports DMIC?

brentlu commented:
Maybe we could test it after this patch got merged.

https://patchwork.kernel.org/project/alsa-devel/patch/6e16433f-1897-46a2-b3ee-e177e7d846b8@moroto.mountain/

plbossart commented:
> Maybe we could test it after this patch got merged.
> 
> https://patchwork.kernel.org/project/alsa-devel/patch/6e16433f-1897-46a2-b3ee-e177e7d846b8@moroto.mountain/

Can you pick this patch on your branch and resubmit @brentlu ?

bardliao commented:
> Maybe we could test it after this patch got merged.
> 
> https://patchwork.kernel.org/project/alsa-devel/patch/6e16433f-1897-46a2-b3ee-e177e7d846b8@moroto.mountain/

I manually tested it on jf-cml-hel-rt5682-05 and it works.

brentlu commented:
RMMOD	snd_soc_acpi_intel_match
rmmod: ERROR: Module snd_soc_acpi_intel_match is in use by: snd_soc_intel_sof_board_helpers

update the sof_remove script :
https://github.com/thesofproject/sof-test/pull/1107

plbossart commented:
SOFCI TEST

plbossart commented:
tests look ok now, merging.
